firmware/modes/programmer/at89/at89.c

```diff
@@ -0,0 +1,654 @@
+     * P3.4     ->      14          RD1                     // Busy
+     * P3.6     <-      16          RG1                     // ctrl (low)
+     * P3.7     <-      17          RE0                     // ctrl (low)
+     */
+    
+    // Set pin direction
+    zif_bits_t dir = {  0,
+                        0b00100000,   // Busy signal (14)
+                        0,
+                        0b10000000,   // p0.7 (32)
+                        0b01111111 }; // p0.{6-0} (33-39)  
+    dir_write(dir);
+    
+    // Set Vdd / GND pinout  
+    set_vdd(vdd);
+    set_gnd(gnd);
+    
+    // Set voltages
+    vdd_val(5); // 5.0 v - 5.2 v
+    
+    // Allocate an empty zifbits struct for reading pin state
+    zif_bits_t response  = { 0, 0, 0, 0, 0 };
+
+    // Base pin setting for reading
+    zif_bits_t signature_base = { 0b00000000,
+                             0b00000001,   // RST (9)
+                             0b00000000,
+                             0b01100000,   // VPP (31), PROG (30)
+                             0b00000000 };
+
+    zif_bits_t signature_clk;
+    
+    // Mask in the address bits to the appropriate pins
+    mask_addr(signature_base, 0x30 + offset);
+
+    // Give the clock on/off states to zif_clock_write(..) and loop 48 cycles
+    clock_write(signature_base, 48);
+
+    // Read the current pin state (to read in the requested byte)
+    zif_read(response);
+
+    // We're done with the byte. Turn off all outputs.
+    zif_write(zbits_null);
+    
+    return zif_to_data(response);
+}
+
+static bool sig_check()
+{
+    if (read_sig(0) != 0x1E || read_sig(1) != 0x51 || read_sig(2) != 0xFF)
+        return false;
+    return true;
+}
+
+static bool blank_check()
+{
+    printf("Performing a blank-check... ");
+    unsigned char data = 0;
+    for (unsigned int addr = 0; addr < 0xFFF; addr++) {
+        printf("%03X", addr);
+        data = read_byte(addr);
+        if (data != 0xFF) {
+            printf("\b\b\bdone. %03X set to byte %02X. Target is not blank.",
+                    addr, data);
+            return false;
+        }
+        printf("\b\b\b");
+    }
+    printf("done. Target is blank.");
+    return true;
+}
+
+static void self_test()
+{
+    printf("Testing first 255 bytes...\r\n");
+    erase();
+    com_println("");
+    for(unsigned int addr = 0; addr < 0xff; addr++) {
+        write(addr, addr);
+        com_println("");
+    }
+    read(0,0xFF);
+    com_println("");
+    printf("Testing last 255 bytes...\r\n");
+    for(unsigned int addr = 0xF00; addr <= 0xFFF; addr++) {
+        write(addr, addr - 0xF00);
+        com_println("");
+    }
+    read(0xF00, 0xFF);
+    printf("\r\nTesting last byte...\r\n");
+    write(0xFFF, 0);
+    com_println("");
+    read(0xFFF, 0);
+    printf("\r\ndone.");
+}
+
+static inline void eval_command(unsigned char * cmd)
+{
+    unsigned char * cmd_t = strtok(cmd, " ");
+    switch (cmd_t[0]) {
+        case 'r':
+        {
+            if (!sig_check()) {

```

Comment:
Its possible to erase the signature using UV light. We should add a way to override this (bonus question: how do we program it back?)

Reply:
Noted. I will add a "force" command to bypass all of these checks.